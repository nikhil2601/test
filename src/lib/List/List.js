import PropTypes from 'prop-types';
import React from 'react';
import styled, { css } from 'styled-components';

import { themeGet } from 'utils/theme';

import { ListItemStyled } from '../ListItem';

const ListStyled = styled.ul`
    position: relative;
    ${({ listStyle, margin, padding }) => css`
        list-style: ${listStyle || 'none'};
        margin: ${margin || 0};
        padding: ${padding || 0};
    `};
    ${ListItemStyled} {
        border-left: ${({ flush }) => flush && 'none'};
        border-right: ${({ flush }) => flush && 'none'};
        &:first-child {
            border-top-left-radius: ${({ flush }) => flush && '0'};
            border-top-right-radius: ${({ flush }) => flush && '0'};
            border-top: ${({ flush }) => flush && 'none'};
        }
        &:last-child {
            border-bottom-left-radius: ${({ flush }) => flush && '0'};
            border-bottom-right-radius: ${({ flush }) => flush && '0'};
            border-bottom: ${({ flush }) => flush && 'none'};
        }
    }
    /**
     * Add dynamic styles
     */
    ${({ backgroundColor, borderRadius }) => css`
        background-color: ${themeGet(`palette.${backgroundColor}.color`, backgroundColor)};
        border-radius: ${borderRadius};
    `}
    /**
     * Add all of the remaining styles from theme
     */
    ${themeGet('List.styles')};
`;

/**
 * Lists are continuous, vertical indexes of text or images.
 *
 * @method      List
 * @constructor
 */
const List = React.forwardRef((props, ref) => {
    const { as, children, ordered, ...rest } = props;
    const listAs = ordered ? 'ol' : 'ul';

    return (
        <ListStyled as={listAs} {...rest} ref={ref}>
            {children}
        </ListStyled>
    );
});

List.propTypes = {
    as: PropTypes.string,
    /**
     * Sets the `background-color` CSS property.
     */
    backgroundColor: PropTypes.string,
    /**
     * Sets the `border-radius` css property.
     */
    borderRadius: PropTypes.string,
    borderStyle: PropTypes.string,
    /**
     * The content of the component.
     */
    children: PropTypes.node,
    /**
     * If `true`, a 1px border is added to the bottom of the list-item.
     */
    divider: PropTypes.bool,
    /**
     * Sets the `list-style` css property.
     */
    listStyle: PropTypes.string,
    /**
     * Sets the `margin` css property.
     */
    margin: PropTypes.string,
    /**
     * Should we render an ordered list `ol` or default to `ul`?
     */
    ordered: PropTypes.bool,
    /**
     * Sets the `padding` css property.
     */
    padding: PropTypes.string,
};

List.defaultProps = {
    as: null,
    backgroundColor: 'transparent',
    borderRadius: null,
    borderStyle: 'initial',
    children: null,
    divider: false,
    listStyle: null,
    margin: null,
    ordered: false,
    padding: null,
};

List.displayName = 'List';

export default List;
