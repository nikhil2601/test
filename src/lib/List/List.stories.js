import React from 'react';

import Box from '../Box';
import List from './List';
import ListItem from '../ListItem';
import { genID } from '../../utils';

export default {
    title: 'List',
    component: List,
};

export const list = () => (
    <Box>
        <List flush>
            {[1, 2, 3, 4].map(item => (
                <ListItem key={genID()} button>
                    Item {item}
                </ListItem>
            ))}
        </List>
    </Box>
);

export const orderedList = () => (
    <Box width="50%">
        <List ordered>
            {[1, 2, 3, 4].map(item => (
                <ListItem key={genID()} button>
                    Item {item}
                </ListItem>
            ))}
        </List>
    </Box>
);

export const bulletList = () => (
    <Box width="50%">
        <List margin="20px" listStyle="disc">
            {[1, 2, 3, 4].map(item => (
                <ListItem key={genID()}>Item {item}</ListItem>
            ))}
        </List>
    </Box>
);

export const listWithDivider = () => (
    <List width="50%" margin="20px">
        {[1, 2, 3, 4].map(item => (
            <ListItem key={genID()} button divider>
                Item {item}
            </ListItem>
        ))}
    </List>
);
