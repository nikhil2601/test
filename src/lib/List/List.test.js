import React from 'react';
import { mount } from 'enzyme';

import List from './index';

describe('<List />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<List />);
    });
    afterEach(() => {
        wrapper.unmount();
    });

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a blank list, without children present', () => {
        expect(wrapper.find('ul')).toHaveLength(1);
        expect(wrapper.find('li')).toHaveLength(0);
    });

    it('should render a list, when children are present', () => {
        wrapper.setProps({ children });
        expect(wrapper.find('ul')).toHaveLength(1);
        expect(wrapper.find('li')).toHaveLength(3);
        expect(wrapper.find('li').at(1)).toHaveText('Item 2');
    });
});

const children = (
    <>
        <li>Item 1</li>
        <li>Item 2</li>
        <li>Item 3</li>
    </>
);
