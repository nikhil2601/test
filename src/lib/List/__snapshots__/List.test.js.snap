// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render correctly 1`] = `
<List
  as={null}
  backgroundColor="transparent"
  borderRadius={null}
  borderStyle="initial"
  divider={false}
  listStyle={null}
  margin={null}
  ordered={false}
  padding={null}
>
  <List__ListStyled
    as="ul"
    backgroundColor="transparent"
    borderRadius={null}
    borderStyle="initial"
    divider={false}
    listStyle={null}
    margin={null}
    padding={null}
  >
    <StyledComponent
      as="ul"
      backgroundColor="transparent"
      borderRadius={null}
      borderStyle="initial"
      divider={false}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "List__ListStyled-sc-57ytho-0",
            "isStatic": false,
            "lastClassName": "kNKJzA",
            "rules": Array [
              "position:relative;",
              [Function],
              ";",
              ".ListItem__ListItemStyled-uh66uc-0",
              "{border-left:",
              [Function],
              ";border-right:",
              [Function],
              ";&:first-child{border-top-left-radius:",
              [Function],
              ";border-top-right-radius:",
              [Function],
              ";border-top:",
              [Function],
              ";}&:last-child{border-bottom-left-radius:",
              [Function],
              ";border-bottom-right-radius:",
              [Function],
              ";border-bottom:",
              [Function],
              ";}}",
              [Function],
              " ",
              [Function],
              ";",
            ],
          },
          "displayName": "List__ListStyled",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "List__ListStyled-sc-57ytho-0",
          "target": "ul",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      listStyle={null}
      margin={null}
      padding={null}
    >
      <ul
        className="List__ListStyled-sc-57ytho-0 kNKJzA"
      />
    </StyledComponent>
  </List__ListStyled>
</List>
`;
